c-dependencies/js-compute-runtime/builtins/subtle-crypto.cpp

```diff
@@ -0,0 +1,407 @@
+  JS::CallArgs args = CallArgsFromVp(argc, vp);
+  if (!args.requireAtLeast(cx, "SubtleCrypto.digest", 2)) {
+    return false;
+  }
+  // 1. Let algorithm be the algorithm parameter passed to the digest() method.
+  auto algorithm = args.get(0);
+
+  // 2 . Let data be the result of getting a copy of the bytes held by the data parameter
+  // passed to the digest() method.
+  size_t length;
+  uint8_t *data = value_to_buffer(cx, args.get(1), "SubtleCrypto#digest: data", &length);
+  if (!data) {
+    return false;
+  }
+
+  std::vector<uint8_t> inputData(data, data + length);
+
+  // 5. Let promise be a new Promise.
+  JS::RootedObject promise(cx, JS::NewPromiseObject(cx, nullptr));
+  if (!promise) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  // 6. Return promise and perform the remaining steps in parallel.
+  args.rval().setObject(*promise);
+
+  // 3. Let normalizedAlgorithm be the result of normalizing an algorithm, with alg set to
+  // algorithm and op set to "digest".
+  // 4. If an error occurred, return a Promise rejected with normalizedAlgorithm.
+  auto normalizedAlgorithmResult = normalizeAlgorithm(cx, algorithm, Operations::Digest);
+  if (normalizedAlgorithmResult.isErr()) {
+    if (!renameExceptionNameToNotSupportedError(cx, promise)) {
+      return RejectPromiseWithPendingError(cx, promise);
+    }
+    return RejectPromiseWithPendingError(cx, promise);
+  }
+  auto normalizedAlgorithm = normalizedAlgorithmResult.unwrap();
+
+  // 7. If the following steps or referenced procedures say to throw an error, reject promise with
+  // the returned error and then terminate the algorithm.
+  // 8. Let result be the result of performing the digest operation specified by normalizedAlgorithm
+  // using algorithm, with data as message.
+
+  size_t digest_length = 0;
+  const EVP_MD *alg;
+  switch (normalizedAlgorithm) {
+    using enum CryptoAlgorithmIdentifier;
+  case SHA_1: {
+    alg = EVP_sha1();
+    digest_length = SHA_DIGEST_LENGTH;
+    break;
+  }
+  case SHA_224: {
+    alg = EVP_sha224();
+    digest_length = SHA224_DIGEST_LENGTH;
+    break;
+  }
+  case SHA_256: {
+    alg = EVP_sha256();
+    digest_length = SHA256_DIGEST_LENGTH;
+    break;
+  }
+  case SHA_384: {
+    alg = EVP_sha384();
+    digest_length = SHA384_DIGEST_LENGTH;
+    break;
+  }
+  case SHA_512: {
+    alg = EVP_sha512();
+    digest_length = SHA512_DIGEST_LENGTH;
+    break;
+  }
+  default: {
+    JS_ReportErrorUTF8(cx, "SubtleCrypto.digest: Operation is not supported");
+    if (!renameExceptionNameToNotSupportedError(cx, promise)) {
+      return RejectPromiseWithPendingError(cx, promise);
+    }
+    return RejectPromiseWithPendingError(cx, promise);
+  }
+  }
+
+  std::vector<uint8_t> output;
+  output.reserve(digest_length);
+  unsigned int size = 0;
+  if (!EVP_Digest(inputData.data(), inputData.size(), static_cast<unsigned char *>(output.data()),
```

Comment:
You don't need to move the input data into the heap -- the `data` and `length` values can be used here instead.
```suggestion
  if (!EVP_Digest(data, length, static_cast<unsigned char *>(output.data()),
```